### PR TITLE
Add MonadUnliftIO instance for InputT

### PR DESCRIPTION
**Pull request by quickdudley**

### CODE REVIEW

**Comment by judah** [comment]
Thank you for the patch.  Unfortunately, I don't think it's feasible for us to support it at this time.  Haskeline is a core dependency of GHC and included as part of its build.  As a result, I tend to be pretty conservative about adding more dependencies to avoid increasing complexity in GHC.  See in particular the recent effect of adding a dependency on `exceptions`: https://gitlab.haskell.org/ghc/ghc/issues/16752

As an alternative, we could consider developing a package with this change as an orphan instance.  If that requires exposing `InputT`'s constructor, maybe we could do so in a `System.Console.Haskeline.Internal` module, with a suitable warning about the API not being as stable.  Let me know if you're interested in taking that on.

**Comment by quickdudley** [comment]
I'm willing to take that on.

**Comment by infinity0** [comment]
An alternative to exposing the constructor is to just define `askUnliftIO` and `withRunInIO` here as top-level functions. The orphan instance definition in another package can then just re-use these functions.

File: System/Console/Haskeline/InputT.hs [hunk 1]
```diff
@@ -65,6 +65,24 @@ instance ( Fail.MonadFail m ) => Fail.MonadFail (InputT m) where
 instance ( MonadFix m ) => MonadFix (InputT m) where
     mfix f = InputT (mfix (unInputT . f))
 
+-- | Run an action in the underlying monad with a function for restoring the
+-- current 'InputT' context.
```
Comment (judah) [inline System/Console/Haskeline/InputT.hs:69]: Could we elaborate the comment a little more, indicating that it's intended for use with other monad transformer stacks? Maybe even call out the unliftio package explicitly as an example?

**Comment by judah** [comment]
Thanks again for this improvement!

**Pull request merged**